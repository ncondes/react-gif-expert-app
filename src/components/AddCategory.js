import { useState } from 'react';
import PropTypes from 'prop-types';


export const AddCategory = ({ setCategories }) => {

  const [ inputValue, setInputValue ] = useState('');

  // handleInputChange
  const handleInputChange = e => setInputValue( e.target.value );
  

  // handleSubmit
  const handleSubmit = e => {
    e.preventDefault();

    if ( inputValue.trim().length > 2 ) {
      // para no tener que recibir la prop de category, podemos usar una funcion flecha para recibir el valor previo que se tenia de estado en categories.
      setCategories( prevCategories => [ inputValue, ...prevCategories ] );
      setInputValue('');
    }
  }

  return (
    <form onSubmit={ handleSubmit }>
      <input
        type='text'
        value={ inputValue }
        onChange={ handleInputChange }
      />
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
}

