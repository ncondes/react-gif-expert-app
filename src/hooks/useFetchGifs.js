import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs'

export const useFetchGifs = ( category ) => {
  
  const [state, setState] = useState({
    data: [],
    loading: true,
  });

  // useEfect hace que la funcion getGifts la cual hace una peticion a una API, solo se haga una vez estableciendo la lista de dependencias ( poniendo los [] )

  // si la categoria llegara a cambiar, use effect no se volveria a reenderizar, pues se establecio [] para que se cargue una unica vez, pero para que useEffect vuelve a realizar un cambio cuando cambia el estado de categories basta con poner como lista de dependencia [ category ] pues es lo que estamos recibiendo como props

  useEffect( () => {

    getGifs( category )
      .then( imgs => {
        // setTimeout para simular una carga mas lenta
        setTimeout(() => {
          setState({
            data: imgs,
            loading: false,
          })
        }, 1500);
      })

  }, [ category ])

  return state;


}
