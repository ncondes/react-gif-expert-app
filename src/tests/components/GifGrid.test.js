import { GifGrid } from '../../components/GifGrid';
import { shallow } from 'enzyme';
import { useFetchGifs } from '../../hooks/useFetchGifs';
// lo que se hace con este mock es fingir cualquier llamada a este archivo y controlar la informacion que eso va a responder
jest.mock('../../hooks/useFetchGifs')

describe('Prueba con GifGrid', () => {

  const category = 'category';

  test('Deberia de hacer match con el snapshot', () => {

    // con mockReturnValue falseamos la data del objeto
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    const wrapper = shallow( <GifGrid
      category={ category }
    /> );
    
    expect( wrapper ).toMatchSnapshot();

  });
  
  test('Debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {
    
    const gifs = [{
      id: 'abc',
      url: 'https://url.com',
      title: 'titulo',
    }];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });
    const wrapper = shallow( <GifGrid
      category={ category }
    /> );

    expect( wrapper ).toMatchSnapshot();
    // mirar que no existe un elemnto p ( loading ) cuando ya se cargo data
    expect( wrapper.find('p').exists() ).toBe( false );
    // como seleccionar componentes / y revisar si la cantidad de items en el grid son iguales a los objetos cargados en gifs
    expect( wrapper.find('GifGridItem').length ).toBe( gifs.length )

  })
  
})