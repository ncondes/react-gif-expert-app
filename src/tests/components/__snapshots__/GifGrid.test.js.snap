// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba con GifGrid Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    ----- category -----
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="abc"
      key="abc"
      title="titulo"
      url="https://url.com"
    />
  </div>
</Fragment>
`;

exports[`Prueba con GifGrid Deberia de hacer match con el snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    ----- category -----
  </h3>
  <p
    className="loading animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
